# Investigating Age-Associated DNA Methylation Changes in _Papio Anubis_

Pipeline for the recreation of a [chapter of Hyeonsoo Jeong's thesis](https://docs.google.com/document/d/1TZFuVnaIoU6e3bUsUtpBu0H6r37STyjj/edit), including: (1) analysis of DNA methylation data from the _anubis_ baboon, (2) development of an aging CpG clock, and (3) further exploratory analysis relating to the relationship between epigenetic aging and early life rearing experience. 

_Project started 03/12/24 as part of a Winter rotation in the Yi lab_. 

### Data
The data were generated by reduced-representation bisulfite sequencing of blood-derived DNA from 140 _Papio anubis_ baboons of a broad age-span. RRBS libraries were constructed using the NuGEN Library Preparation kit and sequenced on an Illumina HiSeq3000 system using 57 bp single-end reads. 

### Newbie Notes
* Installed miniconda using [this tutorial](https://github.com/um-dang/conda_on_the_cluster?tab=readme-ov-file)
* Created a project-specific environment (`dnam`) containing the relevant packages: `bismark`, `trim_galore`, `bowtie2`, `bcftools`, `bedtools`, etc. Remember to run `conda activate dnam` every time you log-in. 
* Remember to move from a login node to a compute node before doing anything demanding (i.e., anything beyond basic file editing and transfer). Either:
  * Submit a job to Slurm using `sbatch`, or
  * Request an interactive computing node using `srun` (e.g., `srun -N 1 -n 4 -p batch --time=4:00:00 --pty bash -i`). This is preferred for compiling or testing code and real-time data exploration. 

## 1. DNA Methylation Data Pre-Processing and Analysis
### Overview of Pipeline 
1. Quality control including read trimming and diversity adaptor removal
2. Bismark read alignment and methylation calling
3. De-duplication
4. Controlling for genetic variation

### Quality Control 
#### Trim Galore!
* Filtered out test-run reads
* Installed `parallel` from `conda-forge` channel
* Created script (`TGjob_parallel.slurm`) to run Trim Galore v.0.4.1. in parallel on the FastA files.
  * Low quality base calls trimmed from 3' end
  * 3' Adaptor sequences removed 
  * Trimmed reads < 20 bp filtered out 
* Ran `sbatch TGjob_parallel.slurm`, with output sent to `~/papanu/TGoutput` (JobID `3435825`).

>[!NOTE]
> The `--rrbs` option (to remove the artificial methylation signal introduced by RRBS sequencing preparation) was not used, because NuGEN method attaches a varying number of nucleotides after each MsPI site. Instead, a diversity trimming step was performed (see below). 

#### NuGEN Diversity Adaptor Trimming
* Downloaded [script provided by NuGEN](https://github.com/nugentechnologies/NuMetRRBS): `nugen_adaptor.py`.
* Created job script (`nugen_job.slurm`) to feed the `trim_galore` output into the python script.
  * Five bases trimmed from 3' end
  * Zero-three bases trimmed at the 5' end (diversity sequences)
  * Any reads not containing 5' MspI site signature removed
* Ran `sbatch nugen_job.slurm`, with output (files with `_trimmed.fq` appended to the filename) sent to `~/papanu/NuGENoutput`.

### Alignment to Reference Genome
`Bismark v 0.14.` was used to concurrently align sequencing reads to the baboon papAnu4 reference genome and perform methylation calling. There are two steps to running Bismark:

1. Genome preparation 
   * On an interactive node, ran `bismark_genome_preparation --verbose ~/papanu/refgenome`
     * Bisulite conversions (all `C`s to `T`s and all `G`s to `A`s) are carried out.
     * `bowtie2` indexer is launched to index the converted genomes in parallel.
2. Read alignment
   * Created and ran job script (`bismark_job.slurm`).
     * Bismark produces an alignment and methylation call output file (`.bam`).

>Question for David: is there a good way to check what resources a job used? I could not get `seff` to work. 

>[!Note]
>From this point on, I have only performed trial runs working with a single file to reduce required computational resources and time.
>
### De-duplication
* On an interactive node, ran `deduplicate_bismark --bam --single 10007_S29_L005_R1_001_trimmed.fq_trimmed_bismark_bt2.bam`
  * Removes alignments to the same position in the genome (ie., if the start and end co-ordinate of the mapped read coincides with any other read, it gets discarded). 
  * These can arise due to excessive PCR amplifcation (i.e. PCR duplicates).

> Question: [felixkrueger](https://felixkrueger.github.io/Bismark/bismark/deduplication/) says "It is important to note that deduplication is not recommended for RRBS, amplicon or other target enrichment-type libraries!"

_Finished here 03/22 (final day of rotation)_
***

### Controlling for Genetic Variation
It is necessary to remove CpG sites that overlap with common genetic variants (genetic polymorphisms of `T` at CpG sites are indistinguishable from bisulfite-converted cytosines). I could not find any description on the best way to do this using [these 100 VCF files cataloging SNPs in baboon genomes](https://zenodo.org/records/2583266); my brainstorming is below. 

- Preparing the variant data
  - Get all genetic variation in the sampled population by merging the VCF files into a single file (possibly using `bcftools merge`).
  - Debating whether to filter for variants that are common (i.e., calculate allele frequency and set a cut-off MAF). This could possibly be done using `bcftools stats`?
- Preparing the RRBS data
  - Perform methylation extraction on the `deduplicated.bam` file via `bismark_methylation_extractor'. This extracts the methylation call for every `C` analysed and outputs `.cov` files for downstream analysis.
  - Extract the CpG sites
- Intersect the RRBS CpG sites with all or common SNPs and see which overlap.
  - Could start by sorting the merged VCF file and the RRBS CpG sites using `bcftools`. 
  - Use `bedtools intersect` to find overlapping CpG sites with common SNPs?
- Could also filter the matching sites based on MAF threshold at this step
- Filter out the determined sites from the RRBS data
 
## 3. Building a CpG Aging Clock

### Elastic Net Regression

### Functional Enrichment

### Identification of Differentially-Methylated CpGs

